File: back_end/109/interview_assessment/small_problems/easy_7/6.rb
=begin

Modify the method from the previous exercise so it ignores non-alphabetic characters when determining whether it should uppercase or lowercase each letter. The non-alphabetic characters should still be included in the return value; they just don't count when toggling the desired case.

input: string
output: a new string
rules: 
- ignore non-letter chars
- include non-letter chars in r.v, but they don't count when determining case
data structure: array, string
pseudo-code: 
- create an empty array
- split string into an array of individual chars
- iterate over chars
- if char is a letter, upcase it
- if char is a non-letter, downcase
algorithm: 
- split string into an array of chars
- iterate using #map
- init. variable lower = ('a'..'z').to_a
- init. variable upper = ('A'..'Z').to_a
- init. variable uppercase = true
- if (&&) LOWER includes char and UPPER includes char
  - if uppercase = true
    - char.upcase
  - if uppercase = false
    - char.downcase
- elsif char is not a letter
    - char
- uppercase = !uppercase
=end

LOWER = ('a'..'z').to_a
UPPER = ('A'..'Z').to_a

def staggered_case(string)
  a = []
  uppercase = true
  string.chars.each do |char|
    if LOWER.include?(char) || UPPER.include?(char)
      if uppercase
        a << char.upcase
      else
        a << char.downcase
      end
    else
      a << char
      uppercase = !uppercase
    end
    uppercase = !uppercase
  end
  a.join
end

p staggered_case('I Love Launch School!')
p staggered_case('ALL CAPS') == 'AlL cApS'
p staggered_case('ignore 77 the 444 numbers') == 'IgNoRe 77 ThE 444 nUmBeRs'

File: exercises/js_small_problems/string_text_processing/6.js
// Write a function that takes a string as an argument and returns that string with a staggered capitalization scheme. Every other character, starting from the first, should be capitalized and should be followed by a lowercase or non-alphabetic character. Non-alphabetic characters should not be changed, but should be counted as characters for determining when to switch between upper and lower case.

/*
input: string of words
output: new string of alternating capitalized characters
rules:
- first letter capitalized
- second character will be lowercased or used as count to determine when to switch case
algorithm:
- declare result to ''
- create an array of all characters in the string
- even indexed spaces will be uppercased
- odd indexed spaces will be lowercased
- return result when all characters have been iterated over
*/

function staggeredCase(string) {
  let result = '';
  let characters = string.split('');

  for (let i = 0; i < characters.length; i += 1) {
    if (i % 2 === 0) {
      result += characters[i].toUpperCase();
    } else {
      result += characters[i].toLowerCase();
    }
  }

  return result;
}

console.log(staggeredCase('I Love Launch School!'));        // "I LoVe lAuNcH ScHoOl!"
console.log(staggeredCase('ALL_CAPS'));                     // "AlL_CaPs"
console.log(staggeredCase('ignore 77 the 4444 numbers'));   // "IgNoRe 77 ThE 4444 nUmBeRs"
